test/functional/plugin/lsp_spec.lua

```diff
@@ -493,6 +493,31 @@ describe('LSP', function()
       eq(0, result.num_attached_after)
     end)
 
+    it('should not re-attach buffer if it was deleted in on_init', function()
```

Comment:
I find the test does not fail without the patch on `vim.lsp.client`.

Reply:
I doubled checked this and it's failing without the client change for me.

Reply:
I see. I also double checked, in my setup it does *not* fail. Is the execution order non-deterministic by any chance?

Actually I don't find the logic of this test and the anticipated order of execution quite clear, we probably meant buffer was gone during `on_init` and expect fail somewhere (but where exactly? some comments will help for elaboration) It looks like `on_init` can be called asynchronously, not necessarily *before* re-attaching. It'd be also a good idea to ensure a certain execution order by assertions.

Reply:
The test assumes that the `on_init` callback will called twice:

1. Before the `vim.lsp.start` returns (`bufnr` should be `nil`)
2. After `client:initialise()` is called. (`bufnr` should be a valid buffer).

Reply:
Thanks, the tests look cleaner after the simplification. A bit off topic, but I was wondering: in which case and why `client:initialize` can be called twice? (it's not a public method, so not supposed to be called by user) `client:initialize` would have been called inside `vim.lsp.start()`. `on_init` is meant to be called only once.

While I think this is a good (though contrived) repro for the bug #28575, what's more impacting is the race condition when the client responds too late, isn't it? Can we possibly come up with a more realistic scenario where #28575 happens?

Reply:
@wookayin `:LspRestart`

Reply:
I meant: calling `client:initialize()` twice in a real-world, not for the tests. If we want to simulate a delay we will have to write a similar code so that the asynchronous callback can be scheduled andexecute, rather than simply calling it from the main thread.

Reply:
I'm not sure what you're suggestion. Do you have an idea for a better way to test this?

Reply:
> I meant: calling client:initialize() twice in a real-world, not for the tests.

I understood the real world scenario as:

- Client starts
- rpc.request("initialize") fires
- Client attaches one or more other buffers
- Server responds to initialize
- on_init fires and deletes a buffer


To mirror that scenario more closely in the test it would be necessary to override the `srv.request` of the `_create_server` and defer the `initialize` response until a second buffer is attached. But I think the current solution is also good enough.

Reply:
I've updated the test, let me know what you think.

Reply:
> or anytime until LSP initialization is done

I don't think this is correct. You can delete the buffer in that window and it should be removed from the `client.attached_buffers` table. See:

https://github.com/neovim/neovim/blob/f79e906442b77cfa4e5a02d51f5dcb2fb186afbc/runtime/lua/vim/lsp.lua#L588
https://github.com/neovim/neovim/blob/f79e906442b77cfa4e5a02d51f5dcb2fb186afbc/runtime/lua/vim/lsp.lua#L678

Reply:
Question: I get you are saying removal of a buffer will invalidate `attached_buffers[bufnr]` (either synchronously or asynchronously, not so sure). But then, to clarify, what's so special about buffer gets wiped out *during* `on_init`? Is this a realistic scenario? I was thinking that the use of `on_init` is an artificial scenario you chose to reproduce the bug, and the race condition could happen as buffer will be removed somewhere else than `on_init`.

Reply:
I'm not sure _why_ someone would be deleting buffers in `on_init`, but based on my read-through of the code, that's the most obvious way the `Invalid buffer id: 44` error from #28575 could happen. It's definitely weird, but it's still a valid usage of the API and therefore should work. This might not end up being the root cause of #28575 but it's still worth fixing if only to narrow the search. The lack of a good repro doesn't leave us with a lot of good options.